### PR TITLE
Move command action user from label to annotation

### LINKED ISSUE #152: Special characters in username causes `/commands` API to fail
# API Call
`/kappnav/resource/commands` with a Cookie containing key/value pair `kappnav-user=IAM%23kueng`

image

# API Response
```
{
  "message": "input-error: Bad Request"
}
```

# Sample stack trace
```
[err] io.kubernetes.client.ApiException: Bad Request
[err] 	at io.kubernetes.client.ApiClient.handleResponse(ApiClient.java:882)
[err] 	at io.kubernetes.client.ApiClient.execute(ApiClient.java:798)
[err] 	at io.kubernetes.client.apis.BatchV1Api.listNamespacedJobWithHttpInfo(BatchV1Api.java:977)
[err] 	at io.kubernetes.client.apis.BatchV1Api.listNamespacedJob(BatchV1Api.java:954)
[err] 	at application.rest.v1.ActionsEndpoint.getCommands(ActionsEndpoint.java:325)
[err] 	at application.rest.v1.ActionsEndpoint$Proxy$_$$_WeldSubclass.getCommands$$super(Unknown Source)
[err] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[err] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[err] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[err] 	at java.lang.reflect.Method.invoke(Method.java:498)
[err] 	at org.jboss.weld.interceptor.proxy.TerminalAroundInvokeInvocationContext.proceedInternal(TerminalAroundInvokeInvocationContext.java:51)
[err] 	at [internal classes]
[err] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[err] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[err] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[err] 	at java.lang.reflect.Method.invoke(Method.java:498)
[err] 	at org.jboss.weld.interceptor.reader.SimpleInterceptorInvocation$SimpleMethodInvocation.invoke(SimpleInterceptorInvocation.java:73)
[err] 	at [internal classes]
[err] 	at application.rest.v1.ActionsEndpoint$Proxy$_$$_WeldSubclass.getCommands(Unknown Source)
[err] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[err] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[err] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[err] 	at java.lang.reflect.Method.invoke(Method.java:498)
[err] 	at com.ibm.ws.jaxrs20.server.LibertyJaxRsServerFactoryBean.performInvocation(LibertyJaxRsServerFactoryBean.java:656)
[err] 	at [internal classes]
[err] 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
[err] 	at com.ibm.websphere.jaxrs.server.IBMRestServlet.service(IBMRestServlet.java:96)
[err] 	at [internal classes]
[err] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[err] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[err] 	at java.lang.Thread.run(Thread.java:748)
```

### PR DESCRIPTION
**Pull request by kinueng**
Fixes kappnav/issues#152

- Move the `kappnav-job-user-id` key/value from labels to annotations in
 support of special characters in usernames.  This change impacts the
APIs that get and submit command actions.

- Add more logging

- Word wrap a few very long code lines

### CODE REVIEW

**Comment by kinueng** [comment]
Example of a job submitted with user `IAM%23kueng`
```
// 20200303215614
// http://localhost:3000/kappnav/resource/commands

{
  "commands": [
    {
      "metadata": {
        "annotations": {
          "kappnav-job-action-text": "Always successful",
          "kappnav-job-user-id": "IAM%23kueng"
        },
        "creationTimestamp": "2020-03-03T21:56:02.000-06:00",
        "labels": {
          "kappnav-job-action-name": "always-successful",
          "kappnav-job-application-name": "stock-trader",
          "kappnav-job-application-namespace": "default",
          "kappnav-job-component-kind": "Deployment",
          "kappnav-job-component-name": "loyalty-level",
          "kappnav-job-component-namespace": "default",
          "kappnav-job-component-sub-kind": "liberty",
          "kappnav-job-type": "command"
        },
        "name": "kappnav-b363f8fd-79ac-4386-8aea-daf50639d3f2",
        "namespace": "kappnav",
        "resourceVersion": "29354489",
        "selfLink": "/apis/batch/v1/namespaces/kappnav/jobs/kappnav-b363f8fd-79ac-4386-8aea-daf50639d3f2",
        "uid": "611d0fd9-1eb5-4f12-8959-e6594098aa7f"
      },
      "spec": {
        "backoffLimit": 4,
        "completions": 1,
        "parallelism": 1,
        "selector": {
          "matchLabels": {
            "controller-uid": "611d0fd9-1eb5-4f12-8959-e6594098aa7f"
          }
        },
        "template": {
          "metadata": {
            "labels": {
              "controller-uid": "611d0fd9-1eb5-4f12-8959-e6594098aa7f",
              "job-name": "kappnav-b363f8fd-79ac-4386-8aea-daf50639d3f2"
            }
          },
          "spec": {
            "containers": [
              {
                "command": [
                  "sleep",
                  "0"
                ],
                "image": "docker.io/kinueng/app-nav-cmds:latest",
                "imagePullPolicy": "Always",
                "name": "kappnav-bdc1aa55-ea70-4c69-9839-2bf67852b766",
                "resources": {
                  
                },
                "terminationMessagePath": "/dev/termination-log",
                "terminationMessagePolicy": "File"
              }
            ],
            "dnsPolicy": "ClusterFirst",
            "restartPolicy": "Never",
            "schedulerName": "default-scheduler",
            "securityContext": {
              "runAsNonRoot": true,
              "runAsUser": 1001
            },
            "serviceAccount": "kappnav-sa",
            "serviceAccountName": "kappnav-sa",
            "terminationGracePeriodSeconds": 30
          }
        }
      },
      "status": {
        "completionTime": "2020-03-03T21:56:13.000-06:00",
        "conditions": [
          {
            "lastProbeTime": "2020-03-03T21:56:13.000-06:00",
            "lastTransitionTime": "2020-03-03T21:56:13.000-06:00",
            "status": "True",
            "type": "Complete"
          }
        ],
        "startTime": "2020-03-03T21:56:03.000-06:00",
        "succeeded": 1
      },
      "kind": "Job"
    }
  ],
  "action-map": {
    "url-actions": [
      {
        "name": "detail",
        "text": "View Detail",
        "text.nls": "action.url.job.detail.text",
        "description": "View Job detail",
        "description.nls": "action.url.job.detail.desc",
        "url-pattern": "${builtin.openshift-admin-console-url}/k8s/ns/${resource.$.metadata.namespace}/jobs/${resource.$.metadata.name}",
        "open-window": "current",
        "menu-item": "false"
      },
      {
        "name": "joblogs",
        "text": "View Job Logs",
        "text.nls": "action.url.job.logs.text",
        "description": "View Job logs",
        "description.nls": "action.url.job.logs.desc",
        "url-pattern": "${builtin.openshift-console-url}/project/${resource.$.metadata.namespace}/browse/pods/${var.podname}?tab=logs",
        "open-window": "tab"
      }
    ],
    "snippets": {
      "name": "function name() { return '{.items[*].metadata.name}'; }"
    },
    "variables": {
      "podname": "${func.kubectlGet(pods,-l,job-name=${resource.$.metadata.name},-o=jsonpath=${snippet.name()})}"
    }
  }
}
```

**Comment by kinueng** [comment]
I built the project with this PR and installed onto the tester's OCP environment.  The problem is resolved and the UI comes up.  I will open a separate issue (Edit: https://github.com/kappnav/issues/issues/154) to address avoiding the UI from catastrophically exploding from an unexpected API response.

File: src/main/java/application/rest/v1/ActionsEndpoint.java [hunk 1]
```diff
@@ -733,27 +752,27 @@ private void setSecurityContextAndServiceAccountName(ApiClient client, V1PodSpec
         spec.setServiceAccountName(config.getkAppNavServiceAccountName()); 
     }
     
-    private Map<String,String> createJobAnnotations(String text) {
+    private Map<String,String> createJobAnnotations(String text, String user) {
         if (text != null && !text.isEmpty()) {
```
Comment (mrglavas) [inline src/main/java/application/rest/v1/ActionsEndpoint.java:758]: This should be used to determine whether the "text" annotation should be set. Setting the user id is independent of that. Need to massage the logic here a bit.

**Review by mrglavas — CHANGES_REQUESTED** [review]
See review comments.

File: src/main/java/application/rest/v1/ActionsEndpoint.java (hunk 1 above)
Comment (kinueng) [reply src/main/java/application/rest/v1/ActionsEndpoint.java:758]: Good catch, @mrglavas.

**Comment by kinueng** [comment]
I tested the new commit f49d15e locally using `kappnav-user=kubeadmin` (legacy user-id label) and `kappnav-user=john%23smith` (new user-id annotation)

**Review by mrglavas — APPROVED** [review]

**Comment by kinueng** [comment]
I retested this PR using AppNav and OCP4.3 without any issues.

**Pull request merged**